### PR TITLE
[fix][client] Release the orphan producers after the primary consumer is closed

### PR DESCRIPTION
**Pull request by poorbarcode**
### Motivation

The producers ["retryLetterProducer", "deadLetterProducer"] will be auto-created by consumers if enabled `DLQ`, but these producers will not close after consumers are closed.

### Modifications

Auto close "retryLetterProducer" and "deadLetterProducer" after the primary consumer is closed

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: 
- https://github.com/poorbarcode/pulsar/pull/79

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 1]
```diff
@@ -1050,7 +1050,19 @@ public CompletableFuture<Void> closeAsync() {
             });
         }
 
-        return closeFuture;
+        return closeFuture.thenCompose(ignore -> {
+            if (retryLetterProducer != null){
+                return retryLetterProducer.closeAsync();
```
Comment (mattisonchao) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:1055]: Should we close them parallel? should we do some special exception handling here?

**Review by Technoboy- — APPROVED** [review]

**Review by lhotari — APPROVED** [review]
LGTM, good catch!

**Review by tisonkun — APPROVED** [review]
Nice catch!

**Review by michaeljmarshall — APPROVED** [review]
Great work! I had tried to fix this a long time ago with #10159 but it was never completed.

**Pull request merged**